vivisect/analysis/aarch64/emulation.py

```diff
@@ -0,0 +1,486 @@
+                    if None not in (base, tbl):
+                        count = len(tbl)
+                        self.switchcases += 1
+                        emu.vw.setVaSetRow('SwitchCases', (op.va, op.va, count))
+
+            if op.iflags & envi.IF_BRANCH:
+                try:
+                    tgt = op.getOperValue(0, emu)
+
+                    if tgt == op.va:
+                        logger.info("0x%x: +++++++++++++++ infinite loop +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++", op.va)
+                        if op.va not in self.infloops:
+                            self.infloops.append(op.va)
+                            if 'InfiniteLoops' not in self.vw.getVaSetNames():
+                                self.vw.addVaSet('InfiniteLoops', (('va', vivisect.VASET_ADDRESS, 'function', vivisect.VASET_STRING)))
+                            self.vw.setVaSetRow('InfiniteLoops', (op.va, self.fva))
+
+                except Exception as e:
+                    self.logAnomaly(emu, self.fva, "0x%x: (%r) ERROR: %s" % (op.va, op, e))
+                    logger.info("0x%x: (%r) ERROR: %s", op.va, op, e)
+
+        except v_exc.BadOpBytes:
+            raise
+
+        except Exception as e:
+            self.logAnomaly(emu, self.fva, "0x%x: (%r) ERROR: %s" % (op.va, op, e))
+            logger.warning("0x%x: (%r)  ERROR: %s", op.va, op, e)
+
+    def posthook(self, emu, op, starteip):
+        if op.opcode == INS_BLX:
+            emu.setFlag(PSR_T_bit, self.last_tmode)
+
+        elif op.opcode == INS_MOVT:
+            val = op.getOperValue(0, emu=emu)
+            emu.vw.setSymHint(op.va, OP_SYMHINT_IDX, val)
+            if emu.isValidPointer(val):
+                emu.vw.addXref(op.va, val, vivisect.REF_PTR)
+
+argnames = {
+    0: ('x0', 0),
+    1: ('x1', 1),
+    2: ('x2', 2),
+    3: ('x3', 3),
+    4: ('x4', 4),
+    5: ('x5', 5),
+    6: ('x6', 6),
+    7: ('x7', 7),
+}
+
+
+def archargname(idx):
+    ret = argnames.get(idx)
+    if ret is None:
+        name = 'arg%d' % idx
+    else:
+        name, idx = ret
+    return name
+
+
+def buildFunctionApi(vw, fva, emu, emumon):
+    argc = 0
+    funcargs = []
+    callconv = vw.getMeta('DefaultCall')
+    undefregs = set(emu.getUninitRegUse())
+
+    for argnum in range(len(argnames), 0, -1):
+        argname, argid = argnames[argnum-1]
+        if argid in undefregs:
+            argc = argnum
+            break
+
+    if callconv == 'a64call':
+        if emumon.stackmax > 0:
+            targc = (emumon.stackmax >> 3) + 6
+            if targc > 40:
+                emumon.logAnomaly(emu, fva, 'Crazy Stack Offset Touched: 0x%.8x' % emumon.stackmax)
+            else:
+                argc = targc
+
+        funcargs = [('int', archargname(i)) for i in range(argc)]
+
+    api = ('int', None, callconv, None, funcargs)
+    vw.setFunctionApi(fva, api)
+    return api
+
+
+def getAllReads(emu):
```

Comment:
Related side note: you and I should talk about the logread/logwrite stuff and maybe an API facelift for it.